include/RAJA/pattern/detail/WorkGroup.hpp

```diff
@@ -0,0 +1,1138 @@
+/*!
+ ******************************************************************************
+ *
+ * \file
+ *
+ * \brief   Header file providing RAJA WorkPool and WorkGroup declarations.
+ *
+ ******************************************************************************
+ */
+
+//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~//
+// Copyright (c) 2016-20, Lawrence Livermore National Security, LLC
+// and RAJA project contributors. See the RAJA/COPYRIGHT file for details.
+//
+// SPDX-License-Identifier: (BSD-3-Clause)
+//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~//
+
+#ifndef RAJA_PATTERN_DETAIL_WorkGroup_HPP
+#define RAJA_PATTERN_DETAIL_WorkGroup_HPP
+
+#include "RAJA/config.hpp"
+
+#include "RAJA/util/Operators.hpp"
+#include "RAJA/util/macros.hpp"
+
+#include "RAJA/policy/WorkGroup.hpp"
+
+#include <type_traits>
+#include <cstddef>
+
+namespace RAJA
+{
+
+namespace detail
+{
+
+template < typename T, typename ... CallArgs >
+void Vtable_move_construct(void* dest, void* src)
+{
+  T* dest_as_T = static_cast<T*>(dest);
+  T* src_as_T = static_cast<T*>(src);
+  new(dest_as_T) T(std::move(*src_as_T));
+}
+
+template < typename T, typename ... CallArgs >
+void Vtable_call(const void* obj, CallArgs... args)
```

Comment:
Shouldn't it be CallArgs&&... for perfect forwarding?

Reply:
The CallArgs are part of the signature of the call function so they have to stay the same as what the user supplied.

Reply:
In the updated version you can see that these are part of the class signature not the method. Do you think I should template the method as well for perfect forwarding?